applications/StructuralMechanicsApplication/custom_utilities/tangent_operator_calculator_utility.h

```diff
@@ -0,0 +1,264 @@
+        const std::size_t num_components = strain_vector_gp.size();
+        // Loop over components of the strain
+        for (std::size_t Component = 0; Component < num_components; Component++)
+        {
+            Vector &perturbed_strain = rValues.GetStrainVector();
+
+            double Perturbation;
+            CalculatePerturbation(perturbed_strain, Component, Perturbation);
+            PerturbateStrainVector(perturbed_strain, strain_vector_gp, Perturbation, Component);
+            IntegratePerturbedStrain(rValues, pConstitutiveLaw);
+
+            Vector &perturbed_integrated_stress = rValues.GetStressVector(); // now integrated
+            const Vector &delta_stress = perturbed_integrated_stress - stress_vector_gp;
+            AssignComponentsToTangentTensor(tangent_tensor, delta_stress, Perturbation, Component);
+
+            // Reset the values to the initial ones
+            noalias(perturbed_strain) = strain_vector_gp;
+            noalias(perturbed_integrated_stress) = stress_vector_gp;
+        }
+    }
+
+    static void CalculatePerturbation(
+        const Vector &StrainVector,
+        const int Component,
+        double &rPerturbation)
+    {
+        double Pert1, Pert2;
+        if (StrainVector[Component] != 0.0)
+        {
+            Pert1 = 1.0e-5 * StrainVector[Component];
+        }
+        else
+        {
+            double MinStrainComp;
+            GetMinAbsValue(StrainVector, MinStrainComp);
+            Pert1 = 1.0e-5 * MinStrainComp;
+        }
+        double MaxStrainComp;
+        GetMaxAbsValue(StrainVector, MaxStrainComp);
+        Pert2 = 1e-10 * MaxStrainComp;
+        rPerturbation = std::max(Pert1, Pert2);
+    }
+
+    static void PerturbateStrainVector(
+        Vector &PerturbedStrainVector,
+        const Vector &strain_vector_gp,
+        const double Perturbation,
+        const int Component)
+    {
+        PerturbedStrainVector = strain_vector_gp;
+        PerturbedStrainVector[Component] += Perturbation;
+    }
+
+    static void IntegratePerturbedStrain(
+        ConstitutiveLaw::Parameters &rValues,
+        ConstitutiveLaw *pConstitutiveLaw)
+    {
+        Flags &cl_options = rValues.GetOptions();
+        // In order to avoid recursivity...
+        cl_options.Set(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR, false);
+
+        pConstitutiveLaw->CalculateMaterialResponseCauchy(rValues);
+    }
+
+    static void GetMaxAbsValue(
+        const Vector &ArrayValues,
+        double &MaxValue)
+    {
+        const int Dim = ArrayValues.size();
+        std::vector<double> non_zero_values;
+
+        for (int i = 1; i < Dim; i++)
+        {
+            if (ArrayValues[i] != 0.0)
+                non_zero_values.push_back(std::abs(ArrayValues[i]));
+        }
+        KRATOS_ERROR_IF(non_zero_values.size() == 0) << "The strain vector is full of 0's..." << std::endl;
+
+        double aux = std::abs(non_zero_values[0]);
+        for (int i = 1; i < non_zero_values.size(); i++)
```

Comment:
usually we use Indextype and ++i for loops (except omp loops, there you have to use int)